pkg/reconciler/apis/permissionclaimlabel/permissionclaimlabel_controller.go

```diff
@@ -0,0 +1,257 @@
+
+	apiBindingsLister apislisters.APIBindingLister
+	listAPIBindings   func(clusterName logicalcluster.Name) ([]*apisv1alpha1.APIBinding, error)
+}
+
+// enqueueAPIBinding enqueues an APIBinding .
+func (c *controller) enqueueAPIBinding(obj interface{}, logSuffix string) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	klog.V(2).Infof("queueing APIBinding %q%s", key, logSuffix)
+	c.queue.Add(key)
+}
+
+// enqueueBindingForResource
+func (c *controller) enqueueBindingForResource(gvr schema.GroupVersionResource, obj interface{}) {
+	// Determine APIBindings that have Permission Claims for the given GVR
+	o, ok := obj.(logicalcluster.Object)
+	if !ok {
+		runtime.HandleError(fmt.Errorf("expected interface: %T but got: %T", o, obj))
+	}
+
+	lc := logicalcluster.From(o)
+
+	apiBindings, err := c.listAPIBindings(lc)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	for _, binding := range apiBindings {
+		for _, pc := range binding.Spec.AcceptedPermissionClaims {
+			if pc.Group == gvr.Group && pc.Resource == gvr.Resource {
+				key := clusters.ToClusterAwareKey(lc, binding.Name)
+				klog.V(2).Infof("queueing APIBinding %q", key)
+				c.queue.Add(key)
+			}
+		}
+	}
+}
+
+// Start starts the controller, which stops when ctx.Done() is closed.
+func (c *controller) Start(ctx context.Context, numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %s controller", controllerName)
+	defer klog.Infof("Shutting down %s controller", controllerName)
+
+	for i := 0; i < numThreads; i++ {
+		go wait.UntilWithContext(ctx, c.startWorker, time.Second)
+	}
+
+	<-ctx.Done()
+}
+
+func (c *controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	if err := c.process(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("%q controller failed to sync %q, err: %w", controllerName, key, err))
+		c.queue.AddRateLimited(key)
+		return true
+	}
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *controller) process(ctx context.Context, key string) error {
+	_, clusterAwareName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		klog.Errorf("invalid key: %q: %v", key, err)
+		return nil
+	}
+	clusterName, name := clusters.SplitClusterAwareKey(clusterAwareName)
+
+	obj, err := c.apiBindingsLister.Get(key)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil // object deleted before we handled it
+		}
+		return err
+	}
+	old := obj
+	obj = obj.DeepCopy()
+
+	reconcileErr := c.reconcile(ctx, obj)
```

Comment:
Here is a slightly better pattern (need to fix the other code in the repo):

```go
var errs []error
if err := c.reconcile(ctx, obj); err != nil {
  errs = append(errs, err)
}

if !equality.Semantic.DeepEqual(old.Status, obj.Status) {
  // ...
  if _, err := c.kcpClusterClient.ApisV1alpha1().APIBindings().Patch(logicalcluster.WithCluster(ctx, clusterName), obj.Name, types.MergePatchType, patchBytes, metav1.PatchOptions{}, "status"); err != nil {
    errs = append(errs, err)
  }
}}

return utilerrors.NewAggregate(errs)

Reply:
Updated to use this pattern